numba/tests/test_types.py

```diff
@@ -217,6 +218,22 @@ def test_list_type_getitem(self):
             self.assertTrue(isinstance(l_int, types.List))
             self.assertTrue(isinstance(l_int[0], type(listty)))
 
+    def test_function_incompatible_templates(self):
+        # issue 4345
+        def func_stub():
+            pass
+
+        def func_stub2():
+            pass
+
+        def ol():
+            pass
+
+        template1 = make_overload_template(func_stub, ol, {}, True)
+        template2 = make_overload_template(func_stub2, ol, {}, True)
+
+        with self.assertRaises(ValueError):
```

Comment:
There's a preference in new code to use the following pattern to check that the exception is the exception expected (`ValueError` and `TypingError` appear 90% of the time and it's very easy to get them from an unexpected place).

```python
with self.assertRaises(ValueError) as raises:
    types.Function((template1, template2))
self.assertIn("incompatible templates:", str(raises.exception))
```